File: gfg/2024_8_august/c++/20_burning_tree.cpp
// TIME COMPEXITY
// O(n)

// SPACE COMPEXITY
// O(n)

// CODE

class Solution {
  public:
  // 1. create mapping
  // 2. find target node
  Node* createMapOfParent(Node* root, int target, map<Node*, Node*> &nodeToParent){
      Node* res = NULL;
      queue<Node*> q;
      q.push(root);
      nodeToParent[root] = NULL;
      while(!q.empty()){
          Node* front = q.front();
          q.pop();
          
          if(front->data == target)
            res = front;
          if(front->left){
              nodeToParent[front->left] = front;
              q.push(front->left);
          }
          if(front->right){
              nodeToParent[front->right] = front;
              q.push(front->right);
          }
      }
      return res;
  }
  
  // 3. burn the tree
  void burnTree(Node* targetNode, map<Node*, Node*> &nodeToParent,int &ans){
    map<Node*, bool> visited;
    queue<Node*> q;
    q.push(targetNode);
    visited[targetNode] = true;
    while(!q.empty()){
        bool flag = 0;
        int size = q.size();
        for(int i = 0; i < size; i++){
            Node* front = q.front();
            q.pop();
            
            // preocess the neighbour nodes left, right, parent
            
            //left
            if(front->left && !visited[front->left]){
                flag = 1;
                q.push(front->left);
                visited[front->left] = 1;
            }
            
            // right
            if(front->right && !visited[front->right]){
                flag = 1;
                q.push(front->right);
                visited[front->right] = 1;
            }
            
            //parent
            if(nodeToParent[front] && !visited[nodeToParent[front]]){
                flag = 1;
                q.push(nodeToParent[front]);
                visited[nodeToParent[front]] = 1;
            }
        }
        if(flag == 1) ans++;
    }
  }
    int minTime(Node* root, int target) 
    {
        int ans = 0;
        map<Node*, Node*> nodeToParent;
        Node* targetNode = createMapOfParent(root,target,nodeToParent);
        burnTree(targetNode,nodeToParent,ans);
        return ans;
    }
};

File: gfg/2024_8_august/python/20_burning_tree.py
# CODE

from collections import defaultdict, deque

class Solution:
    def minTime(self, root, target):
        adj = defaultdict(list)
        x = None

        def buildAdjList(root):
            nonlocal x
            if root is None:
                return

            if root.left:
                adj[root].append(root.left)
                adj[root.left].append(root)

            if root.right:
                adj[root].append(root.right)
                adj[root.right].append(root)

            if root.data == target:
                x = root

            buildAdjList(root.left)
            buildAdjList(root.right)

        buildAdjList(root)

        # BFS to calculate the maximum time to burn the entire tree
        q = deque([(x, 0)])
        vis = set()
        res = 0

        while q:
            u, time = q.popleft()
            res = max(res, time)
            vis.add(u)

            for neighbor in adj[u]:
                if neighbor not in vis:
                    vis.add(neighbor)
                    q.append((neighbor, time + 1))

        return res